Week4/LESSON_PLAN.md

```diff
@@ -0,0 +1,275 @@
+and explain the concepts better in the class.
+
+## Topics
+
+1. Embedded vs Normalised data modeling
+2. Advanced MongoDB operations
+3. Pagination
+4. Indexes in MongoDB
+5. Transactions in MongoDB
+6. SQL vs NoSQL databases
+
+## 1. Embedded vs Normalised data modeling
+
+### Explanation
+
+1. Embedded means that information is stored in a sub-object in a collection
+2. Normalised means that information is stored in different collection but are referencing each other
+
+### Example
+
+Let's have a look at a bug logging database.
+
+#### The embedded way
+
+```js
+const bugs = [
+  {
+    timestamp: new Date(),
+    page: "about",
+    stacktrace: "", // removed for clarity
+    reporter: {
+      email: "rob@thebugcreator.com",
+      name: "Rob",
+    },
+  },
+];
+```
+
+#### The normalised way
+
+```js
+const accounts = [
+  {
+    id: ObjectId("507f191e810c19729de86032"),
+    email: "rob@thebugcreator.com",
+    name: "Rob",
+  },
+];
+
+const bugs = [
+  {
+    timestamp: new Date(),
+    page: "about",
+    stacktrace: "", // removed for clarity
+    reporterId: ObjectId("507f191e810c19729de86032"),
+  },
+];
+```
+
+### Exercise
+
+Discuss the differences and what the advantages/disadvantages are to each approach. For example:
+
+Embedded allows for faster querying.
+Normalized allows for less data duplication.
+
+### Essence
+
+There are advantages to both approaches and in the wild you will have to decide which to use every time.
+
+## 2. Advanced MongoDB operations
+
+### Explanation
+
+1. The `sort` command allows you to sort data that you get back from your query.
+2. The `limit` command allows you to limit how many items you get back from your query.
+3. The `aggregate` command allows the combination and calculation of data in one or more collections.
+
+### Example
+
+Let's assume a log database with the following information:
+
+```js
+const bugs = [
+  {
+    timestamp: new Date('2000-06-07T11:24:00'),
+    page: "about",
+    stacktrace: "", // removed for clarity
+    reporterId: ObjectId("507f191e810c19729de86032"),
+  }, {
+    timestamp: new Date('2000-06-06T12:23:00'),
+    page: "about",
+    stacktrace: "", // removed for clarity
+    reporterId: ObjectId("507f191e810c19729de86032"),
+  }. {
+    timestamp: new Date('2000-06-08T12:33:00'),
+    page: "contact",
+    stacktrace: "", // removed for clarity
+    reporterId: ObjectId("e810507f191de86032c19729"),
+  }, {
+    timestamp: new Date('2000-06-06T12:34:00'),
+    page: "home",
+    stacktrace: "", // removed for clarity
+    reporterId: ObjectId("e810507f191de86032c19729"),
+  }
+];
+```
+
+#### Sort
+
+If we want to sort the find query on the timestamp to find the latest bugs we can run the following query:
+
+```js
+client.db("logging").collection("bugs").find().sort({ timestamp: -1 });
+```
+
+#### Limit
+
+The above query will give you back all of the bugs which is not great, as such we can use the limit command as follows:
```

Comment:
I don't agree. One could want to see all the bugs with a certain timestamp. 

`LIMIT` clause is especially useful when top-10 or top-5 elements are sought (e.g. Top-10 countries with the most population, top-5 bugs that occurred at midnight)